### PR TITLE
[flang][OpenMP] Rewrite `omp.loop` to semantically equivalent ops

### PR DESCRIPTION
**Pull request by ergawy**
Introduces a new conversion pass that rewrites `omp.loop` ops to their
semantically equivalent op nests bases on the surrounding/binding
context of the `loop` op. Not all forms of `omp.loop` are supported yet.
See `isLoopConversionSupported` for more info on which forms are supported.

### CODE REVIEW

File: flang/include/flang/Optimizer/OpenMP/Passes.td
Comment (tblah) [inline flang/include/flang/Optimizer/OpenMP/Passes.td:54]: Is there any chance we could have loop wrapper ops nested inside of other operations some day?

I'm not saying this is a real case - just an example: say we used a parallel do inside of an omp.private copy region to copy each element of an array in parallel.

I imagine this is a no, I just wanted to check.

File: flang/include/flang/Optimizer/OpenMP/Passes.td
Comment (tblah) [inline flang/include/flang/Optimizer/OpenMP/Passes.td:62]: I'm not sure how important these dependent dialects are, but it must at least depend on mlir::func::FuncDialect as this is a mlir::func::FuncOp pass?

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (tblah) [inline flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:160]: You could use the loc in `func` here

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (tblah) [inline flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:158]: This might just be because the pass is WIP, but it looks like you are trying to do a partial conversion here: the conversion pattern could just return that the pattern did not match if `!isLoopConversionSupported(...)`.

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (tblah) [inline flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:96]: It reads a bit weird to me constructing a new vector and assigning that. I have to stop and remember that the result of the constructor is an xvalue and so it won't have to copy the data. Maybe this would work and be easier to read?
```c++
auto &privateSyms = loopOp.getPrivateSyms();
parallelClauseOps.privateSyms.assign(privateSyms.begin(), privateSyms.end());
```

File: flang/test/Lower/OpenMP/generic-loop-rewriting.f90
Comment (tblah) [inline flang/test/Lower/OpenMP/generic-loop-rewriting.f90:1]: When this is ready for review, it would be good to see an mlir-to-mlir test which runs only the new pass. This makes it easier to review exactly what the pass is doing. It is up to you if you keep this lowering test as well.

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (ergawy) [reply flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp]: A partial conversion ignores unknown ops (e.g. from unregistered dialects) but does not allow ops explicitly marked illegal to return a match failure. In that sense, I am doing a full conversion: we are converting all illegal `omp.loop` ops. If there is an unsupported `omp.loop` op, I was just marking it as legal and leaving the `todo` to be emitted later on by the OpenMP to LLVM IR translation logic.

Comment (tblah) [reply flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:158]: Okay my understanding must have been incorrect.

**Review by mjklemm — APPROVED** [review]
LGTM, except small typo

File: flang/include/flang/Optimizer/OpenMP/Passes.td
Comment (ergawy) [reply flang/include/flang/Optimizer/OpenMP/Passes.td:60]: That's a good quesiton and I am not sure about the answer. But my intuitiion is that even if we need a loop to initialize a `firstprivate` variable, then this loop will exists in a function of its own; for example, a copy constructor or an overload of the assignment operator which means the loop is still inside a function.

@mjklemm please correct me if I am wrong here.

Comment (mjklemm) [reply flang/include/flang/Optimizer/OpenMP/Passes.td:60]: I think that's correct.  In that case, the generated code would likely call a runtime function for intrinsic assignment.  Maybe the best way is to clearly document this assumption, so that when it turns out wrong in the future we can easily update it.

**Review by tblah — APPROVED** [review]
Thanks for the updates and explanations

**Comment by kiranchandramohan** [comment]
I have a nit comment about the name. `GenericLoopConversionPass` seems a bit too Generic. 

Is this implementing parts of https://discourse.llvm.org/t/rfc-representing-combined-composite-constructs-in-the-openmp-dialect/76986? If so a link to the discourse post would be useful.

**Comment by ergawy** [comment]
> I have a nit comment about the name. `GenericLoopConversionPass` seems a bit too Generic.

The term `GenericLoop` is adapted from clang: `OMPGenericLoopDirective` (see for example `emitTargetTeamsGenericLoopRegionAsParallel` and `emitTargetTeamsGenericLoopRegionAsDistribute`). Since I am planning to handle all the cases relevant to the `omp.loop` in this new pass, I just went this name.
 
> Is this implementing parts of https://discourse.llvm.org/t/rfc-representing-combined-composite-constructs-in-the-openmp-dialect/76986? If so a link to the discourse post would be useful.

Done.

File: flang/include/flang/Optimizer/OpenMP/Passes.td
Comment (ergawy) [reply flang/include/flang/Optimizer/OpenMP/Passes.td:60]: Done.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:1577]:
```suggestion
  genEntryBlock(builder, args, sectionsOp.getRegion());
```

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:1582]:
```suggestion
    genEntryBlock(builder, args, op->getRegion(0));
```

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:79]:
```suggestion
      if (mlir::isa_and_present<mlir::omp::TargetOp>(teamsOp->getParentOp()))
```

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:84]:
```suggestion
      if (mlir::isa_and_present<mlir::omp::TargetOp>(parallelOp->getParentOp()))
```

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:146]:
```suggestion
    if (func.isDeclaration())
      return;
```

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:163]: This could be removed after emitting errors directly from `isLoopConversionSupported()`

File: flang/test/Transforms/generic-loop-rewriting.mlir
Comment (skatrak) [reply flang/test/Transforms/generic-loop-rewriting.mlir:38]: I wouldn't say we don't have to worry about it, but rather there are some updates needed in other places. We'd have to make sure to enforce host-evaluation of loop bounds for `target teams loop` or equivalent construct nests in the `omp.target` verifier, update Flang lowering accordingly and then this pass seems to me that it should work unmodified for the tests in this PR.

However, what's expected to happen in this case (or equivalent Fortran)?
```c++
#pragma omp target teams
{
  #pragma omp loop
  for (...) { ... }

  #pragma omp loop
  for (...) { ... }
}
```

In that case, loop bounds wouldn't be host-evaluated if this pass introduces two `distribute parallel for` nests, resulting in something similar to what is being produced by this pass right now. If these values being shared causes issues, perhaps we would need to  introduce a firstprivate clause into the created `omp.parallel` ops.

**Review by skatrak — COMMENTED** [review]
Thank you Kareem for this work, and sorry for the delay getting to it! I have a few minor suggestions and comments.

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (ergawy) [reply flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:54]: I did this intentionally to defer error reporting to `OpenMPToLLVMIRTranslation.cpp`. An unsupported loop will not result in a conversion failure. Rather the loop will be marked as "legal" and be allowed to continue down the pipeline. The advantage of doing this is to:
1. Use already existing error reporting in LLVM conversion.
2. During MLIR to MLIR conversion testing, we don't have to worry about handling errors from unsupported loop forms.

See also this [comment](https://github.com/llvm/llvm-project/pull/115443#discussion_r1846055696).

File: flang/include/flang/Optimizer/OpenMP/Passes.td
Comment (ergawy) [reply flang/include/flang/Optimizer/OpenMP/Passes.td:74]: My understanding from the code is a bit different, see this [comment](https://github.com/llvm/llvm-project/pull/115443#discussion_r1846014692).

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (kiranchandramohan) [inline flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:13]: Do we need all these headers?

File: flang/include/flang/Common/OpenMP-utils.h
Comment (ergawy) [reply flang/include/flang/Common/OpenMP-utils.h:19]: There is one other instance of OpenMP stuff in `OpenMP-features.h` which has a single function. I'd rather use a namespace rather than adding `OpenMP` to all the utils in the file. Changed to `Fortran::common::openmp` though.

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (kiranchandramohan) [reply flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:13]: Same for all the MLIR headers below.

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (kiranchandramohan) [inline flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:32]: I thought this pass would be in the MLIR directory. Is there any reason for it to be not there?

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (ergawy) [reply flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp]: Most of them can go for sure. Updated to include the minimum required.

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (ergawy) [reply flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:28]: For consistency with other OpenMP passes, nothing more. I agree with you though, we could probably move pretty much everything under `flang/lib/Optimizer/OpenMP/` to the dialect directory under `mlir`. But I think this should be done in a separate PR in one go and keep things consistent here.

Let me know if you disagree.

Comment (kiranchandramohan) [reply flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:28]: Handling in a separate PR is fine with me.

>  I agree with you though, we could probably move pretty much everything under flang/lib/Optimizer/OpenMP/ to the dialect directory under mlir

I am not fully sure about this. Generic transformations should go to MLIR. But there could be some transformations that are Flang specific where for ease of lowering we lower in one way but immediately correct in a pass to the right form.

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (skatrak) [reply flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp:50]: Ah, I see. I thought this pass would fail because it emits an error. Then, it seems like we should emit a warning instead below.

The issue I see with that approach is that, whenever full support for the loop directive is implemented, no `omp.loop` ops are expected to reach translation to LLVM IR, since they would have been replaced by this pass. Hence, it seems to me that this is the right place to emit these errors now.

At the moment, translation to LLVM IR it will just abort saying "not yet implemented: omp.loop", which I think is not enough information. We could call `checkImplementationStatus` for it, and add cases for this operation to produce the proper per-clause TODO error messages, but I feel it wouldn't be the right place for it because the implementation is going to be in this pass.

File: flang/include/flang/Optimizer/OpenMP/Passes.td
Comment (skatrak) [reply flang/include/flang/Optimizer/OpenMP/Passes.td:74]: Yes, what you say in that comment is consistent with the fact that the `dependentDialects` property gets translated into an implementation for `getDependentDialects()`. What I'm saying is that, per the docs, this must return those dialects holding operations that can be created by the pass.

My understanding of this is that, if you're reading some other dialect's operation in a pass, it must have been created by something else before, so that dialect is already loaded. If you're creating an operation, its dialect might not have been loaded, so you need to make sure of it.

Comment (ergawy) [reply flang/include/flang/Optimizer/OpenMP/Passes.td]: Removed the `FuncDialect`

File: flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp
Comment (ergawy) [reply flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp]: Added error messages and relevant tests.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:45]: Nit: For consistency, it looks like `Fortran::common::omp` would be a better name for the namespace in the end.

**Review by skatrak — APPROVED** [review]
Thank you for working on my comments, this LGTM. Please wait for @kiranchandramohan's approval before merging.

**Review by kiranchandramohan — APPROVED** [review]
LG

**Pull request merged**
